scipy/optimize/tests/test_nonlin.py

```diff
@@ -319,6 +320,33 @@ def test_krylov(self):
         self._check(nonlin.KrylovJacobian, 20, 2, False, inner_m=10)
         self._check(nonlin.KrylovJacobian, 20, 2, True, inner_m=10)
 
+    def _check_autojac(self, A, b):
+        def func(x):
+            return A.dot(x) - b
+
+        def jac(v):
+            return A
+
+        sol = nonlin.nonlin_solve(func, np.zeros(b.shape[0]), jac, maxiter=2,
+                                  f_tol=1e-6, line_search=None, verbose=0)
+        assert_(np.allclose(A @ sol, b, atol=1e-6))
+        # test jac input as array -- not a function
+        sol = nonlin.nonlin_solve(func, np.zeros(b.shape[0]), A, maxiter=2,
+                                  f_tol=1e-6, line_search=None, verbose=0)
+        assert_(np.allclose(A @ sol, b, atol=1e-6))
```

Comment:
I suspect you're just copying the testing approach from other parts of this module, but I'd like to ask that we try to use `np.assert_allclose` instead of `assert_(np.allclose...` for at least two reasons:

1) The error messages when `assert_allclose()` fails can be quite a bit more helpful than `assert_` and `allclose()` because we can get degree and extent of mismatch printed out automatically.
2) It is perfectly safe to use plain `assert` with `pytest` now, and we have a slight preference for that approach moving forward.

Obviously, these are just mechanical/surface level comments, but since you haven't changed any old tests and just added new ones, it does seem like we might be moving forward from non-expert perspective.

Reply:
Good idea...  I changed all the `allclose` tests in this module.  If you prefer I only do it for the new tests I can pull the other changes back out.

Reply:
I think you misunderstood, it should be `assert_allclose()`, not `assert np.allclose()`, per point 1 above. There were two problems, but the solution to both of them is to just use `assert_allclose()`.

I'd prefer if you only made the change for the new tests for now, because the changes you made to the old tests were also not what we want, so let's just keep it scoped for now. Also, have to be careful on the tolerance default differences between `assert_allclose` and `allclose`.

Reply:
Ahh... you mean `np.testing.assert_allclose`?  There doesn't seem to be any `np.assert_allclose`. That's why I thought it must mean `assert np.allclose`.  I didn't see the `np.testing.assert_allclose`. Should be OK now unless I misunderstood again.